server/src/main/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhase.java

```diff
@@ -123,6 +135,70 @@ protected SearchPhase getNextPhase(SearchPhaseResults<CanMatchResponse> results,
         return new GroupShardsIterator<>(sortShards(shardsIts, results.minAndMaxes, fieldSort.order()));
     }
 
+    /**
+     * Iterate over the shard iterator and look for overlapping rollup indices. If rollup indices exist,
+     * the method includes the shards from the optimal rollup index that can match the query and ignores
+     * the shards from other overlapping indices.
+     *
+     * @return A {@link FixedBitSet} with the shards of the optimal indices.
+     */
+    private FixedBitSet chooseOptimalRollupShards(GroupShardsIterator<SearchShardIterator> shardsIts, FixedBitSet possibleMatches,
+                                                  ClusterState clusterState, SearchSourceBuilder source) {
+        // Map with key the index that will be replaced by an optimal rollup index. The optimal index
+        // will be its replacement. In the end all indices contains in the keys will be ignored.
+        Map<String, Tuple<String, RollupIndexMetadata>> indexReplacements = new HashMap<>(shardsIts.size());
+
+        int i = 0;
+        for (SearchShardIterator iter : shardsIts) {
+            if (possibleMatches.get(i++)) {
+                String indexName = iter.shardId().getIndexName();
+                Map<String, String> idxMetadata = clusterState.getMetadata().index(indexName).getCustomData(RollupIndexMetadata.TYPE);
+                IndexAbstraction originalIndex = clusterState.getMetadata().getIndicesLookup().get(indexName);
+
+                // If index is not a member of a datastream it will not be replaced
+                if (idxMetadata == null || originalIndex.getParentDataStream() == null
+                    || indexReplacements.containsKey(indexName)) {
+                    continue;
+                }
+
+                try {
+                    String sourceIndex = idxMetadata.get(RollupIndexMetadata.SOURCE_INDEX_NAME_META_FIELD);
+                    RollupIndexMetadata rollupIndexMetadata = RollupIndexMetadata.parseMetadataXContent(
+                        idxMetadata.get(RollupIndexMetadata.ROLLUP_META_FIELD));
+                    if (indexReplacements.containsKey(sourceIndex)) {
+                        // Retrieve the previously optimal index and compare with the current index
+                        Tuple<String, RollupIndexMetadata> previousOptimalIndex = indexReplacements.get(sourceIndex);
+                        Map<String, RollupIndexMetadata> rollupGroup = Map.of(
+                            indexName, rollupIndexMetadata,
+                            previousOptimalIndex.v1(), previousOptimalIndex.v2());
+                        String newOptimalIndex = RollupShardDecider.findOptimalIndex(rollupGroup, source.aggregations());
+                        Tuple<String, RollupIndexMetadata> optimalTuple = Tuple.tuple(newOptimalIndex, rollupGroup.get(newOptimalIndex));
+                        // replace original index with the new optimal index
```

Comment:
would it be clearer to build these rollup groups and then find the optimalTuple. it seems like this is being run every time a new index that is part of a rollup is found or am I reading this wrong

Reply:
I thought that a single pass approach would be simpler. As every shard is processed we check if its index is the optimal one.  If not we replace its index with the optimal index.

Alternatively, a double pass approach would require building the rollup groups (a map of lists) in a first pass and then iterate over the lists of rollup indices to compute the optimal indices. 

Performance-wise both approaches are similar as each index is compared against the previously optimal index. If you think I am missing something here, I am happy to change this part.

Reply:
should be fine as is, was just wondering if you thought  it would be clearer the other way.